core/src/mast/node_fingerprint.rs

```diff
@@ -0,0 +1,190 @@
+use alloc::{collections::BTreeMap, vec::Vec};
+
+use miden_crypto::hash::{
+    blake::{Blake3Digest, Blake3_256},
+    rpo::RpoDigest,
+    Digest,
+};
+
+use crate::{
+    mast::{DecoratorId, MastForest, MastNode, MastNodeId},
+    Operation,
+};
+
+// MAST NODE EQUALITY
+// ================================================================================================
+
+pub type DecoratorFingerprint = Blake3Digest<32>;
+
+/// Represents the hash used to test for equality between [`MastNode`]s.
+///
+/// The decorator root will be `None` if and only if there are no decorators attached to the node,
+/// and all children have no decorator roots (meaning that there are no decorators in all the
+/// descendants).
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct MastNodeFingerprint {
+    mast_root: RpoDigest,
+    decorator_root: Option<DecoratorFingerprint>,
+}
+
+// ------------------------------------------------------------------------------------------------
+/// Constructors
+impl MastNodeFingerprint {
+    /// Creates a new [`MastNodeFingerprint`] from the given MAST root with an empty decorator root.
+    pub fn new(mast_root: RpoDigest) -> Self {
+        Self { mast_root, decorator_root: None }
+    }
+
+    /// Creates a new [`MastNodeFingerprint`] from the given MAST root and the given
+    /// [`DecoratorFingerprint`].
+    pub fn with_decorator_root(mast_root: RpoDigest, decorator_root: DecoratorFingerprint) -> Self {
+        Self {
+            mast_root,
+            decorator_root: Some(decorator_root),
+        }
+    }
+
+    /// Creates a [`MastNodeFingerprint`] from a [`MastNode`].
+    ///
+    /// The `hash_by_node_id` map must contain all children of the node for efficient lookup of
+    /// their fingerprints.
+    pub fn from_mast_node(
+        forest: &MastForest,
+        hash_by_node_id: &BTreeMap<MastNodeId, MastNodeFingerprint>,
+        node: &MastNode,
+    ) -> MastNodeFingerprint {
```

Comment:
Perhaps we should return an error here instead of panicking when `hash_by_node_id` does not contain the id of a child node?

Reply:
Since this is a part of the public API now, I think that would be ideal. But also, I'm fine leaving it as is for now and creating an issue to do this later on (would be a "good first issue"). I would mention the panic in the doc comments though (just make it explicit).